server/komodo-rest/src/main/java/org/komodo/rest/service/KomodoMetadataService.java

```diff
@@ -3576,17 +3581,97 @@ public Response publishVirtualization(final @Context HttpHeaders headers, final
         }
         UnitOfWork uow = null;
         try {
-            Repository repo = this.kengine.getDefaultRepository();
-            uow = createTransaction(principal, "publish", true); //$NON-NLS-1$
+            uow = createTransaction(principal, "publish-init", true); //$NON-NLS-1$
             Vdb vdb = findVdb(uow, attributes.getName());
+            if (vdb == null) {
+                //
+                // We don't have a vdb so maybe we have a dataservice instead
+                // Find the dataservice's vdb to publish.
+                //
+                Dataservice dataservice = findDataservice(uow, attributes.getName());
+                if (dataservice == null) {
+                    return createErrorResponse(Status.NOT_FOUND, mediaTypes, RelationalMessages.Error.VDB_NOT_FOUND);
+                }
+
+                vdb = dataservice.getServiceVdb(uow);
+            }
+
             if (vdb == null) {
                 return createErrorResponse(Status.NOT_FOUND, mediaTypes, RelationalMessages.Error.VDB_NOT_FOUND);
             }
-            // the properties in this class can be exposed for user input
-            PublishConfiguration config = new PublishConfiguration();
-            config.setVDB(vdb);
-            BuildStatus status = this.openshiftClient.publishVirtualization(uow, config);
-            return commit(uow, mediaTypes, entityFactory.createBuildStatus(uow, repo, status, uriInfo.getBaseUri()));
+
+            KomodoStatusObject status = new KomodoStatusObject();
+            status.addAttribute("Publishing", "Operation initiated");
+
+            final String vdbPath = vdb.getAbsolutePath();
+            final AuthToken token = getAuthenticationToken();
+
+            //
+            // This publishing could take 30+ seconds to return hence the need to
+            // check it then return before the http request actually times out.
+            //
+            // Waits up to 10 seconds to see if the publishing returns (and probably
+            // has failed!) then returns an indication that things are underway.
+            //
+            ExecutorService threadService = Executors.newFixedThreadPool(1);
```

Comment:
Why would publishing timeout in 30 secs? I did not see this behavior when I was using swagger API. Timeout IMO is typically set by the calling client on their expectations. The publishing had sync part and async part, you are adding another layer on top of it which I do think is correct.

If for some reason you need to return in 10secs, we could have modified the publishing API to be totally async with one more state, maybe like "submitted" and encapsulate this behavior to return the call immediately (not wait for pom generation, initial build)

Reply:
1. Publishing does not timeout in 30 seconds, the http request does. Please see:
* https://seancoates.com/blogs/api-gateway-timeout-workaround/
* https://stackoverflow.com/questions/13582409/http-client-timeout-and-server-timeout

2. As noted in the commit messages, the publishing operation up to adding-to-queue (when the new thread kicks in) takes 28-45 seconds on my computer. It is possible that your computer has been running on the low end of this range so have avoided it.

3. This timeout IS set by the calling client, ie. the browser and 30 seconds is plenty long enough to keep a TCP connection open.

4. The threading is necessary as the sync part is taking too long and returning a 504 Gateway Timeout error when it shoud return a 200 response.

5. Whether this is another layer is irrelevant. It is necessary for any operation that on the system will take a long time. For example, if exporting or connection creation in the future take longer than 30 seconds then they too will require a threading modification.

6. The operation did NOT need to return in 10 seconds and, as stated in the commit messages, provides a means of ensuring that the if the publishing returns prematurely due to an error, that error can be immediately returned via the direct response rather than awaiting the polling. The 10 seconds can be reduced but upon testing seemed a useful middle-of-the-road value.

7. Your suggestion of further modification of the publishing API to be totally async is a good one. However, within the scope of this set of commits, I felt it was one step too far. Certainly, a follow-on to introduce additional states would be beneficial, especially as I have noted the following (but not actually witnessed it):
 * User initiates publish
 * Publish request goes to komodo and kicks of the 'publishing build' thread
 * The request waits 10 seconds for any error then returns a response of publishing initiated
 * The UI returns a successful message to say publishing under way and the icon starts to spin
 * Meanwhile, the building of the content is slow and takes 45 seconds rather than ~32 and at that point
 * The UI polls komodo and notes there is nothing in the queue so changes the publishing icon back to "Not Published" (not spinning)
 * The UI polls again, finds something on the queue and changes the icon back to "publishing" (spinning)

To reiterate I have not seen this in testing as my build takes about ~32 seconds which means its possible that the polling never kicks in before the publishing service joins the queue. Therefore, putting everything on a permanent queue would be better to avoid this potential.

Reply:
Can you create a JIRA for this for full async mode? I will see if I can fix that. What is your polling frequency?

Reply:
see [TEIIDTOOLS-400](https://issues.jboss.org/browse/TEIIDTOOLS-400).

Polling occurs every 5 seconds.